File: App/Resources/AlgorithmDetails/randomguesssort/randomguesssort.rb
def sort(arr)
  n = arr.length
  return if n <= 1
  loops = Array.new(n, 0)
  loop do
    is_sorted = true
    (0...n - 1).each do |i|
      a = arr[loops[i]]
      b = arr[loops[i + 1]]
      if a < b || (a == b && loops[i] < loops[i + 1])
        next
      end
      is_sorted = false
      break
    end
    break if is_sorted
    (0...n).each do |pos|
      if loops[pos] < n - 1
        loops[pos] += 1
        break
      end
      loops[pos] = 0
    end
  end

  mapped = loops.map { |i| arr[i] }
  (0...n).each do |i|
    arr[i] = mapped[i]
  end
end

array = [0, 39, 21, 62, 14]
sort(array)
p array

File: App/Resources/AlgorithmDetails/randomguesssort/randomguesssort.cs
using System;

public class RandomGuessSort
{
  public static Random r = new Random();

  public static void Sort(int[] arr)
  {
    int n = arr.Length;
    int[] loops = new int[n];
    while (true)
    {
      bool isSorted = true;
      for (int i = 0; i < n - 1; i++)
      {
        int a = arr[loops[i]];
        int b = arr[loops[i + 1]];
        if (a < b || (a == b && loops[i] < loops[i + 1]))
        {
          continue;
        }
        isSorted = false;
        break;
      }
      if (isSorted)
      {
        break;
      }
      for (int pos = 0; pos < n; pos++)
      {
        loops[pos] = r.Next(n);
      }
    }

    int[] mapped = new int[n];
    for (int i = 0; i < n; i++)
    {
      mapped[i] = arr[loops[i]];
    }
    for (int i = 0; i < n; i++)
    {
      arr[i] = mapped[i];
    }
  }

  public static void Main(String[] args)
  {
    int[] array = { 0, 39, 21, 62, 14 };
    Sort(array);
    string result = "[" + String.Join(", ", array) + "]";
    Console.WriteLine(result);
  }
}